README
SIFO Overview
=============
[SIFO] is a well-proben open source PHP5 framework currently running on several 
sites, from small installations to large websites with multiple servers.

SIFO is free and very easy to use. Contains a lot of libraries and classes that
allow you to get your projects running in a breeze.

You can download the code from both [Github] [GH] or [Google Code] [GC]

 [SIFO]: http://sifo.me
 [GH]: https://github.com/alombarte/SIFO "Visit the Github SIFO repository"
 [GC]: http://code.google.com/p/sifo/ "Visit the Google Code SIFO repository"




Architecture
------------
In this section we will describe a little bit the architecture of SIFO.

### MVC pattern ###
SIFO uses the MVC pattern to sepparate your project in 3 different areas:

* Models: Contain the bussiness logic. Where all the data is taken from (e.g: Queries to database)
* Views: HTML templates with a little bit of logic such as iterations. It's just the presentation of your web. [Smarty] by default.
* Controllers: The ones that put it all together with any necessary logic.

  [Smarty]: http://www.smarty.net

If you never used this pattern before, you should start now if you don't want to
become insane once your applications start to grow. Don't feel overwhelmed, it's easy.



### Multiple environments ###
Via a single file, the mighty `config/domains.config.php`, you set-up your application.
In this file you declare the domains your application will listen to and set the
configuration for each one of them. For instance, if you want to code the webpage
**http://sifo.me** you might want to set at least 2 domains: The real production 
domain *sifo.me* and a fake one for local development *sifo.local*

In the `config/domains.config.php` you can set up things like:

* Show/hide the debug (`devel` flag)
* Languages accepted by your application
* Credentials to several services, init commands, master/slave configurations (e.g: mysql)
* Related hostnames, error reporting and other specific PHP tasks.

*Portion of the file*:

	$config['sifo.me'] = array(
		'devel' => false, // No debug in production.
		'instance' => 'sifoweb',
		'language' => 'en_US',
		'database' => array(
			'profile' => 'PRODUCTION' // Master/slave profiled taken from db_profiles.config.php.
			'db_driver' => 'mysqli',
			'db_host' => '127.0.0.1',
			'db_user' => 'root',
			'db_password' => 'root',
			'db_name' => 'mydatabase',
			'db_init_commands' => array( 'SET NAMES utf8' )
		),
		/*  REDIS syntax:
		'database' => array(
			'database' => array(
			'host'     => '127.0.0.1',
			'port'     => 6379,
			'database' => 0
		*/
		),
		'php_ini_sets' => array( // Empty array if don't want changes.
			// Log errors to 'logs' folder:
			'log_errors' => 'On',
			'error_log' => ROOT_PATH . '/logs/sifoweb_errors.log',
		),
		'libraries_profile' => 'stable_libraries'
		);

		$config['sifo.local'] = $config['sifo.me']; // Take all the config from production
		// Redefine some vars for local domain:
		$config['sifo.local']['devel'] = true;
		$config['sifo.local']['static_host'] = 'http://static.sifo.local';
		$config['libraries_profile'] = 'unstable_libraries'


### Routing ###
When you type an URL there is a mapping between what you are typing in the address bar
and the controller that should reply to that request. Is what we call the *router* and
is located under `config/router.config.php`. If your application has urls translated
you can also use the `router_xx_XX.config.php` files, where the xx_XX is the language
and country code of your app. SIFO supports as much languages as you need.

### Powerful Debug ###
The SIFO debug contains a lot of interesting stuff to debug your application:

* Benchmark: Full decomposition and analysis of times of execution per file and method
* Parameters that every Controller received
* Variables every template received (smarty)
* Queries launched to the database
	* Automatic commenting of the queries
	* Nice print of the final SQL launched
	* Different coloring for write/read operations, with times
	* Slow query alert
	* Printed recordset returned in tabular data
	* Host, database, and replica the query went in
* Data stored in Session (and *kill session*)
* Data stored in user Cookies
* Cache control (names of the keys, remaining TTL...)
* Sphinx Debug
* Cache invalidation
	* AUTOMATIC: While browsing
	* Manual: By flag



### Multilanguage ###
Almost every SIFO programmer has multi-lingual projects, because it is very easy.
We even have translation tools based on config files or in database, so you can
pass the URLs to your non-programmers colleagues the URLs for proper translation.

All the smarty templates come with the i18n plugins that allow you to write your
files in a single language. Example: If we had an application with three possible
languages, we could have in the template something like:

	<h1>{t 1=$username}Hello %1, welcome back!{/t}</h1>

	This code would produce a different output depending on the current active language:

	<h1>Hello John, welcome back!</h1>
	<h1>Hola Juan, bienvenido de nuevo!</h1>
	<h1>Hola Joan, benvingut de nou!</h1>

Your application languages are set on the `config/domains.config.php` file.

Sometimes we've been asked, yes, there is support for non-latin alphabets. We do
have several installations in Russian, Arabic, Japanese, Chinese, Hebrew... so yes,
and also design with RTL text.	

### Lazy loading ###
The less classes are in-memory, the better. We stick to the principle of loading
the classe only if they are going to be used. The configuration file `config/libraries.config.php`
defines what classes are loaded per execution, and what versions of the libraries
do you use. You can even create several profiles of libraries: maybe you want to
play with the latest unstable library in local, while keeping the stable one in production.

Example:

	$config['classes_always_preloaded'] = array(
		'Exceptions',
		'Registry',
		'Cache',
		'CantLiveWithoutThisClass'
	);

	// Contains all the libraries available.
	$config['default'] = array(
		'mylibrary' => 'Mylibrary-1.5-stable'
	);
	$config['bleeding_edge_insane'] = array(
		'mylibrary' => 'Mylibrary-2.0alpha-unstable'
	);

At execution time you can always load any needed application via the 
`$this->getClass( 'ClassName' )` function in any point of the code, or if it's
going to be loaded for sure include it in a protected var:

	// Include the Filter class before the Controller is actually executed:
	protected $include_classes = array( 'Filter' );

### Multiple databases flavour ###
From Key-values like Redis to Oracle, Mysql, Postgres, Firebird or even SQLite.
You can choose to use Mysql PDO or any of the ADODB drivers. For redis, Predis
is an excellent solution.

### Static files sepparation ###
You never know what is going to happen. Sepparate your static files in a different
host is always a win-win. 

### Lots of helpful classes ###
Many projects share the same needs. Do you need to send nice emails? Geolocalise
the users? Facebook/Youtube/Twitter integration? Quick searches using Sphinx? Upload
files to Amazon S3? Do not reinvent the wheel, invest your time in something else.

### Extendable ###
What would be the framework if you couldn't add our own classes and libraries?
There is room for entire libraries, simple classes or *helpers*. Before coding
something from scratch, ensure there is not something done already.


### Command Line support ###
Yes, you like to code your cron jobs in PHP, so do we. Take profit of your existing
models. Create your scripts extending from the Command Line.
The CLI tool lets you define the supported parameters of the script and automatically
builds the help for you. If you are interested in receiveing email reporting is also
already done.

### And many more ###
I am just tired of writing, is late night and I think you read enough. If you arrived
here it's time to get a little bit deeper yourself guided by the documentation.




CHANGELOG.md
# Changelog

## [3.0.2](https://github.com/sifophp/sifo/tree/3.0.2) (2022-05-27)

[Full Changelog](https://github.com/sifophp/sifo/compare/3.0.1...3.0.2)

**Merged pull requests:**

- Add symfony security checker action to pipeline [\#142](https://github.com/sifophp/SIFO/pull/142) ([kpicaza](https://github.com/kpicaza))
- Add easy way to instantiate domains [\#141](https://github.com/sifophp/SIFO/pull/141) ([kpicaza](https://github.com/kpicaza))
- allow to put argv via $\_SERVER [\#140](https://github.com/sifophp/SIFO/pull/140) ([kpicaza](https://github.com/kpicaza))
- AD-231 Add secure and http\_only parameters support for cookies [\#139](https://github.com/sifophp/SIFO/pull/139) ([destebang](https://github.com/destebang))

## [3.0.1](https://github.com/sifophp/sifo/tree/3.0.1) (2022-03-14)

[Full Changelog](https://github.com/sifophp/sifo/compare/3.0.0...3.0.1)

**Merged pull requests:**

- fix: default parameters for query debug call [\#138](https://github.com/sifophp/SIFO/pull/138) ([kpicaza](https://github.com/kpicaza))

## [3.0.0](https://github.com/sifophp/sifo/tree/3.0.0) (2022-02-25)

[Full Changelog](https://github.com/sifophp/sifo/compare/v2.32.0...3.0.0)

**Closed issues:**

- Unit testing for Sifo classes [\#57](https://github.com/sifophp/SIFO/issues/57)
- Change getClass\(\) to support the new autoloader and inheritance [\#40](https://github.com/sifophp/SIFO/issues/40)
- Adapt existing code to PSR-4 [\#39](https://github.com/sifophp/SIFO/issues/39)
- Create new PSR directory structure with Composer [\#38](https://github.com/sifophp/SIFO/issues/38)
- Composer dependencies and repositories definition [\#36](https://github.com/sifophp/SIFO/issues/36)
- Analysis of migration needed in Sifo 2.xx instances [\#32](https://github.com/sifophp/SIFO/issues/32)

**Merged pull requests:**

- Prepare release [\#137](https://github.com/sifophp/SIFO/pull/137) ([kpicaza](https://github.com/kpicaza))
- Trigger deprecation warning when fixing query params [\#136](https://github.com/sifophp/SIFO/pull/136) ([kpicaza](https://github.com/kpicaza))
- Remove Bash scripts folder [\#135](https://github.com/sifophp/SIFO/pull/135) ([kpicaza](https://github.com/kpicaza))
- Fix: improve query fixer regex by Edu Fabra [\#134](https://github.com/sifophp/SIFO/pull/134) ([kpicaza](https://github.com/kpicaza))
- Use maintained version of PHPThumb library [\#133](https://github.com/sifophp/SIFO/pull/133) ([kpicaza](https://github.com/kpicaza))
- Remove unused & deprecated google translate feature [\#132](https://github.com/sifophp/SIFO/pull/132) ([kpicaza](https://github.com/kpicaza))
- MDI-1466: support official adodb5 library [\#131](https://github.com/sifophp/SIFO/pull/131) ([kpicaza](https://github.com/kpicaza))
- MDI-1435 Removed mysql deprecated files [\#129](https://github.com/sifophp/SIFO/pull/129) ([seergiue](https://github.com/seergiue))
- MDI-1433: drop support for adodb-sybase [\#128](https://github.com/sifophp/SIFO/pull/128) ([kpicaza](https://github.com/kpicaza))
- MDI-1445: drop support for adodb-encrypt-mcrypt [\#127](https://github.com/sifophp/SIFO/pull/127) ([kpicaza](https://github.com/kpicaza))
- MDI-1439: set addo cache dir default as if safe mode disabled [\#126](https://github.com/sifophp/SIFO/pull/126) ([kpicaza](https://github.com/kpicaza))
- MDI-1438: drop support for adodb.mysql [\#125](https://github.com/sifophp/SIFO/pull/125) ([kpicaza](https://github.com/kpicaza))
- MDI-1437: drop support for adodb.mssql [\#124](https://github.com/sifophp/SIFO/pull/124) ([kpicaza](https://github.com/kpicaza))
- MDI-1436: drop support for adodb.ibase [\#123](https://github.com/sifophp/SIFO/pull/123) ([kpicaza](https://github.com/kpicaza))
- MDI-1440 Removed deprecated get/set magic\_quotes methods [\#122](https://github.com/sifophp/SIFO/pull/122) ([seergiue](https://github.com/seergiue))
- MDI-1442 Removed deprecated "is\_dst" parameter [\#121](https://github.com/sifophp/SIFO/pull/121) ([seergiue](https://github.com/seergiue))
- MDI-1443 Removed global var variable [\#120](https://github.com/sifophp/SIFO/pull/120) ([seergiue](https://github.com/seergiue))
- MDI-1444 Removed deprecated session\_register\(\) function [\#119](https://github.com/sifophp/SIFO/pull/119) ([seergiue](https://github.com/seergiue))
- MDI-1441 Removed/replaced deprecated unset\($this\) [\#118](https://github.com/sifophp/SIFO/pull/118) ([seergiue](https://github.com/seergiue))
- MDI-1432 Delete class using extension 'fbsql' because it is removed since PHP 5.3 [\#117](https://github.com/sifophp/SIFO/pull/117) ([antoniorova](https://github.com/antoniorova))
- MDI-1431 Fix WARNING: INI directive 'magic\_quotes\_sybase' is deprected since PHP 5.3 [\#116](https://github.com/sifophp/SIFO/pull/116) ([antoniorova](https://github.com/antoniorova))
- MDI-1428 Deleted all functions that are using get\_magic\_quotes\_gpc [\#115](https://github.com/sifophp/SIFO/pull/115) ([antoniorova](https://github.com/antoniorova))
- MDI-1425 Fix warning: Function each\(\) is deprecated since PHP 7.2 [\#114](https://github.com/sifophp/SIFO/pull/114) ([antoniorova](https://github.com/antoniorova))
- MDI-1434 Replaced $php\_errormsg with error\_get\_last\(\) [\#113](https://github.com/sifophp/SIFO/pull/113) ([seergiue](https://github.com/seergiue))
- MDI-1430 removed/replaced deprecated OCI functions [\#112](https://github.com/sifophp/SIFO/pull/112) ([seergiue](https://github.com/seergiue))
- MDI-1429 Rename constructor class name to \_\_construct [\#111](https://github.com/sifophp/SIFO/pull/111) ([seergiue](https://github.com/seergiue))
- MDI-1427 Remove dl\(\) function as it is deprecated [\#110](https://github.com/sifophp/SIFO/pull/110) ([seergiue](https://github.com/seergiue))
- MDI-1426 Replace deprecated function split with explode [\#109](https://github.com/sifophp/SIFO/pull/109) ([seergiue](https://github.com/seergiue))

constants.php
<?php

define('ROOT_PATH', __DIR__);
define('ADODB_DIR', __DIR__ . '/src/adodb5');
define('ADODB_BAD_RS', '<p>Bad $rs in %s. Connection or SQL invalid. Try using $connection->debug=true;</p>');
define('ADODB_FETCH_NUM', 1);
define('ADODB_FETCH_ASSOC', 2);
define('ADODB_FETCH_BOTH', 3);
define('ADODB_FETCH_DEFAULT', 0);
define('ADODB_OPT_HIGH', 2);
define('ADODB_OPT_LOW', 1);
define('ADODB_ASSOC_CASE',0);
define('ADODB_PHPVER',74000);
define('ADODB_EXTENSION', '.php');
define('ADODB_TABLE_REGEX','([]0-9a-z_\:\"\`\.\@\[-]*)');


define('PEAR_ERROR_RETURN', 1);
define('SMARTY_RESOURCE_CHAR_SET', 'utf-8');

define('IFX_SCROLL',1);

define('TIMESTAMP_FIRST_YEAR',0);

define("DB_ERROR", -1);
define("DB_ERROR_SYNTAX", -2);
define("DB_ERROR_CONSTRAINT", -3);
define("DB_ERROR_NOT_FOUND", -4);
define("DB_ERROR_ALREADY_EXISTS", -5);
define("DB_ERROR_UNSUPPORTED", -6);
define("DB_ERROR_MISMATCH", -7);
define("DB_ERROR_INVALID", -8);
define("DB_ERROR_NOT_CAPABLE", -9);
define("DB_ERROR_TRUNCATED", -10);
define("DB_ERROR_INVALID_NUMBER", -11);
define("DB_ERROR_INVALID_DATE", -12);
define("DB_ERROR_DIVZERO", -13);
define("DB_ERROR_NODBSELECTED", -14);
define("DB_ERROR_CANNOT_CREATE", -15);
define("DB_ERROR_CANNOT_DELETE", -16);
define("DB_ERROR_CANNOT_DROP", -17);
define("DB_ERROR_NOSUCHTABLE", -18);
define("DB_ERROR_NOSUCHFIELD", -19);
define("DB_ERROR_NEED_MORE_DATA", -20);
define("DB_ERROR_NOT_LOCKED", -21);
define("DB_ERROR_VALUE_COUNT_ON_ROW", -22);
define("DB_ERROR_INVALID_DSN", -23);
define("DB_ERROR_CONNECT_FAILED", -24);
define("DB_ERROR_EXTENSION_NOT_FOUND", -25);
define("DB_ERROR_NOSUCHDB", -30);
define("DB_ERROR_ACCESS_VIOLATION", -26);
define("DB_ERROR_DEADLOCK", -27);
define("DB_ERROR_STATEMENT_TIMEOUT", -28);
define("DB_ERROR_SERIALIZATION_FAILURE", -29);
define('DB_FETCHMODE_ASSOC', 2);
